{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6e68a69f",
   "metadata": {},
   "source": [
    "## Parse out main text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e3c5018",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys \n",
    "from lxml import etree\n",
    "from unidecode import unidecode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "74711f61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/data/xzuo/Projects/dissertation/dataset_URL_parser/paper_oriented', '/usr/local/anaconda3/lib/python39.zip', '/usr/local/anaconda3/lib/python3.9', '/usr/local/anaconda3/lib/python3.9/lib-dynload', '', '/home/xzuo/.local/lib/python3.9/site-packages', '/usr/local/anaconda3/lib/python3.9/site-packages', '/usr/local/anaconda3/lib/python3.9/site-packages/locket-0.2.1-py3.9.egg', '/usr/local/anaconda3/lib/python3.9/site-packages/IPython/extensions', '/home/xzuo/.ipython', '/data/xzuo/.conda/envs/dissertation/lib/python38.zip', '/data/xzuo/.conda/envs/dissertation/lib/python3.8', '/data/xzuo/.conda/envs/dissertation/lib/python3.8/lib-dynload', '/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages', '/data/xzuo/.conda/envs/dissertation/lib/python38.zip', '/data/xzuo/.conda/envs/dissertation/lib/python3.8', '/data/xzuo/.conda/envs/dissertation/lib/python3.8/lib-dynload', '/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages']\n"
     ]
    }
   ],
   "source": [
    "# Update env and pkg paths.\n",
    "\n",
    "sys.path.extend(['/data/xzuo/.conda/envs/dissertation/lib/python38.zip', '/data/xzuo/.conda/envs/dissertation/lib/python3.8', '/data/xzuo/.conda/envs/dissertation/lib/python3.8/lib-dynload', '/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages'])\n",
    "print(sys.path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6d304da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pubmed_parser as pp\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f5d9d2b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get pmid from LitCovid.\n",
    "\n",
    "id_list_path = \"./litcovid_pmid/litcovid_16052022.tsv\"\n",
    "csv_directory_path = \"./litcovid_pmcid_lists\"\n",
    "download_directory_path =\"./litcovid_pmc_fulltext/xml\"\n",
    "\n",
    "current_filelist = os.listdir(download_directory_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcd0c557",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove inaccessible files.\n",
    "\n",
    "inaccessible_articles = []\n",
    "\n",
    "for i, xml_name in enumerate(current_filelist):\n",
    "    pmc_xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    try:\n",
    "        pmc_tree = etree.parse(pmc_xml_path)\n",
    "    except:\n",
    "        # print(\"Remove file %s.\" % xml_name)\n",
    "        inaccessible_articles.append(xml_name)\n",
    "        current_filelist.remove(xml_name)\n",
    "\n",
    "for i, xml_name in enumerate(current_filelist):\n",
    "    pmc_xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    pmc_tree = etree.parse(pmc_xml_path)\n",
    "    meta = pmc_tree.find(\".//article-meta\")\n",
    "    if meta is None:\n",
    "        inaccessible_articles.append(xml_name)\n",
    "        current_filelist.remove(xml_name)\n",
    "#         print(\"File %s is not accessible.\" % xml_name)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "print(\"There are %d articles not accessible.\" % len(inaccessible_articles))\n",
    "\n",
    "for i, xml_name in enumerate(inaccessible_articles):\n",
    "    pmc_xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    if os.path.exists(pmc_xml_path):\n",
    "        os.remove(pmc_xml_path)\n",
    "    else:\n",
    "        print(\"Error: File %s does not exist.\" % xml_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b2e3cfde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>main text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34776649</td>\n",
       "      <td>8572530</td>\n",
       "      <td>Subjektives Befinden und Arbeitsfähigkeit nach...</td>\n",
       "      <td>Hintergrund  Seit dem 29.01.2021 wurde der C...</td>\n",
       "      <td>Gegen SARS-CoV‑2 wurden Impfstoffe mit völlig ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>34098882</td>\n",
       "      <td>8182346</td>\n",
       "      <td>Web-based, rapid and contactless management of...</td>\n",
       "      <td>Background  During the SARS-CoV-2 pandemic a...</td>\n",
       "      <td>From the end of February 2020 onwards, many pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34541298</td>\n",
       "      <td>8441688</td>\n",
       "      <td>Molecular Level Dissection of Critical Spike M...</td>\n",
       "      <td>Abstract  SARS‐CoV‐2 virus during its spread ...</td>\n",
       "      <td>Severe Acute Respiratory Syndrome Corona Virus...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32302815</td>\n",
       "      <td>7151272</td>\n",
       "      <td>Psychological status of surgical staff during ...</td>\n",
       "      <td>Highlight     •  The mental health problems o...</td>\n",
       "      <td>Dear editor,\\nIn the outbreak of novel coronav...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>34295171</td>\n",
       "      <td>8291382</td>\n",
       "      <td>The Immunogenicity of Hyaluronic Fillers and I...</td>\n",
       "      <td>Abstract  Hyaluronic acid (HA) is a glycosami...</td>\n",
       "      <td>Hyaluronic acid (HA) is a glycosaminoglycan, a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61889</th>\n",
       "      <td>34115405</td>\n",
       "      <td>8420322</td>\n",
       "      <td>The presence of SARS‐CoV‐2 virus in semen samp...</td>\n",
       "      <td>Abstract  The SARS‐CoV‐2 set off a pandemic i...</td>\n",
       "      <td>Coronavirus (SARS‐CoV‐2) induced a pandemic al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61890</th>\n",
       "      <td>35029196</td>\n",
       "      <td>8757937</td>\n",
       "      <td>Does influenza vaccination help reduce inciden...</td>\n",
       "      <td>Abstract  To facilitate the understanding of ...</td>\n",
       "      <td>Corona virus disease 2019 (COVID-19) was decla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61891</th>\n",
       "      <td>35414326</td>\n",
       "      <td>9115793</td>\n",
       "      <td>Myocarditis following COVID-19 vaccination: in...</td>\n",
       "      <td>ABSTRACT   Introduction  Vaccines have demons...</td>\n",
       "      <td>The concept of introducing substrate resemblin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61892</th>\n",
       "      <td>34655522</td>\n",
       "      <td>8514195</td>\n",
       "      <td>Safety and immunogenicity of CpG 1018 and alum...</td>\n",
       "      <td>Background  MVC-COV1901, a recombinant prote...</td>\n",
       "      <td>\\nEvidence before this study\\nTo understand th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61893</th>\n",
       "      <td>32376573</td>\n",
       "      <td>7167319</td>\n",
       "      <td>COVID-19临床分型与MSCT容积扫描间的相关性</td>\n",
       "      <td>目的  探讨新型冠状病毒肺炎(COVID-19)不同的临床分型与多层螺旋计算机体层摄影术...</td>\n",
       "      <td>\\nEvidence before this study\\nTo understand th...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61894 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pmid    pmcid                                              title  \\\n",
       "0      34776649  8572530  Subjektives Befinden und Arbeitsfähigkeit nach...   \n",
       "1      34098882  8182346  Web-based, rapid and contactless management of...   \n",
       "2      34541298  8441688  Molecular Level Dissection of Critical Spike M...   \n",
       "3      32302815  7151272  Psychological status of surgical staff during ...   \n",
       "4      34295171  8291382  The Immunogenicity of Hyaluronic Fillers and I...   \n",
       "...         ...      ...                                                ...   \n",
       "61889  34115405  8420322  The presence of SARS‐CoV‐2 virus in semen samp...   \n",
       "61890  35029196  8757937  Does influenza vaccination help reduce inciden...   \n",
       "61891  35414326  9115793  Myocarditis following COVID-19 vaccination: in...   \n",
       "61892  34655522  8514195  Safety and immunogenicity of CpG 1018 and alum...   \n",
       "61893  32376573  7167319                         COVID-19临床分型与MSCT容积扫描间的相关性   \n",
       "\n",
       "                                                abstract  \\\n",
       "0        Hintergrund  Seit dem 29.01.2021 wurde der C...   \n",
       "1        Background  During the SARS-CoV-2 pandemic a...   \n",
       "2       Abstract  SARS‐CoV‐2 virus during its spread ...   \n",
       "3       Highlight     •  The mental health problems o...   \n",
       "4       Abstract  Hyaluronic acid (HA) is a glycosami...   \n",
       "...                                                  ...   \n",
       "61889   Abstract  The SARS‐CoV‐2 set off a pandemic i...   \n",
       "61890   Abstract  To facilitate the understanding of ...   \n",
       "61891   ABSTRACT   Introduction  Vaccines have demons...   \n",
       "61892    Background  MVC-COV1901, a recombinant prote...   \n",
       "61893    目的  探讨新型冠状病毒肺炎(COVID-19)不同的临床分型与多层螺旋计算机体层摄影术...   \n",
       "\n",
       "                                               main text  \n",
       "0      Gegen SARS-CoV‑2 wurden Impfstoffe mit völlig ...  \n",
       "1      From the end of February 2020 onwards, many pa...  \n",
       "2      Severe Acute Respiratory Syndrome Corona Virus...  \n",
       "3      Dear editor,\\nIn the outbreak of novel coronav...  \n",
       "4      Hyaluronic acid (HA) is a glycosaminoglycan, a...  \n",
       "...                                                  ...  \n",
       "61889  Coronavirus (SARS‐CoV‐2) induced a pandemic al...  \n",
       "61890  Corona virus disease 2019 (COVID-19) was decla...  \n",
       "61891  The concept of introducing substrate resemblin...  \n",
       "61892  \\nEvidence before this study\\nTo understand th...  \n",
       "61893  \\nEvidence before this study\\nTo understand th...  \n",
       "\n",
       "[61894 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Parse out text.\n",
    "articles = []\n",
    "\n",
    "for i, xml_name in enumerate(current_filelist):\n",
    "    pmc_xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    pmc_tree = etree.parse(pmc_xml_path)\n",
    "    meta = pmc_tree.find(\".//article-meta\")\n",
    "    try:\n",
    "        metadata = pp.parse_pubmed_xml(pmc_xml_path)\n",
    "        full_text = pp.parse_pubmed_paragraph(pmc_xml_path, all_paragraph=True)\n",
    "        paras = []\n",
    "        for para in full_text:\n",
    "            paras.append(para['text'].strip())\n",
    "            main_text = \"\\n\".join(paras)\n",
    "        pmc_article = {\"pmid\": metadata['pmid'], \"pmcid\": metadata['pmc'], \n",
    "                       \"title\": metadata['full_title'], \"abstract\": metadata['abstract'], \n",
    "                       \"main text\": main_text}\n",
    "        articles.append(pmc_article)\n",
    "    except:\n",
    "        os.remove(pmc_xml_path)\n",
    "        print(\"File %s is not accessible. Removed.\" % xml_name)\n",
    "\n",
    "df = pd.DataFrame(articles)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10c1d471",
   "metadata": {},
   "source": [
    "## General Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c907fd01",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysbd\n",
    "import unidecode\n",
    "from nltk.tokenize import word_tokenize"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "750e637e",
   "metadata": {},
   "source": [
    "### Search dataset description sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "00b6669d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Searched 1000 full-text articles.\n"
     ]
    }
   ],
   "source": [
    "seg = pysbd.Segmenter(language=\"en\", clean=False)\n",
    "pattern = re.compile(r'((?:train|test|validation|testing|trainings?)\\s*(?:set|data))|(benchmarks?)|(data\\s*(?:set|base)s?)|(corpus)|(corpora)|(tree\\s*bank)|(collections?)')\n",
    "\n",
    "for i, article in enumerate(articles):\n",
    "    all_text = article['title'] + \".\\n\" + article['abstract'] + \"\\n\" + article['main text']\n",
    "    article['sentences'] = seg.segment(all_text)\n",
    "    article['has_pattern'] = []\n",
    "    for j, sent in enumerate(article['sentences']):\n",
    "        if pattern.search(sent):\n",
    "            article['has_pattern'].append(sent)\n",
    "    if i != 0 and i % 1000 == 0:\n",
    "        print(\"Searched %d full-text articles.\" % i)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "14ce9946",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>main text</th>\n",
       "      <th>sentences</th>\n",
       "      <th>has_pattern</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34847386</td>\n",
       "      <td>8609674</td>\n",
       "      <td>Uncertainty-aware convolutional neural network...</td>\n",
       "      <td>Deep learning (DL) has shown great success in...</td>\n",
       "      <td>Novel Corona Virus disease 2019 (COVID-19) has...</td>\n",
       "      <td>[Uncertainty-aware convolutional neural networ...</td>\n",
       "      <td>[The proposed method is evaluated on the three...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>34976572</td>\n",
       "      <td>8675556</td>\n",
       "      <td>Detecting SARS-CoV-2 From Chest X-Ray Using Ar...</td>\n",
       "      <td>Chest radiographs (X-rays) combined with Deep...</td>\n",
       "      <td>The Severe Acute Respiratory Syndrome Coronavi...</td>\n",
       "      <td>[Detecting SARS-CoV-2 From Chest X-Ray Using A...</td>\n",
       "      <td>[However, questions remain regarding the accur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34352699</td>\n",
       "      <td>8249716</td>\n",
       "      <td>Conventional and microfluidic methods for airb...</td>\n",
       "      <td>Graphical abstract      With the COVID-19 pan...</td>\n",
       "      <td>Throughout history, infectious diseases have p...</td>\n",
       "      <td>[Conventional and microfluidic methods for air...</td>\n",
       "      <td>[Creating targeted and effective interventions...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32836689</td>\n",
       "      <td>7236753</td>\n",
       "      <td>Competitive pricing of substitute products und...</td>\n",
       "      <td>There has been an increased interest in optim...</td>\n",
       "      <td>Retailers frequently rely on suppliers that ar...</td>\n",
       "      <td>[Competitive pricing of substitute products un...</td>\n",
       "      <td>[Section 3 is devoted to benchmark modeling of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>34070422</td>\n",
       "      <td>8226892</td>\n",
       "      <td>Shall I Work with Them? A Knowledge Graph-Base...</td>\n",
       "      <td>We consider the prediction of future research...</td>\n",
       "      <td>The development of knowledge graph-based appro...</td>\n",
       "      <td>[Shall I Work with Them? , A Knowledge Graph-B...</td>\n",
       "      <td>[We benchmark the proposed approach against cl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>34139910</td>\n",
       "      <td>8216038</td>\n",
       "      <td>Protecting Privacy and Transforming COVID-19 C...</td>\n",
       "      <td>Objectives  Federal open-data initiatives th...</td>\n",
       "      <td>Federal open-data initiatives that promote inc...</td>\n",
       "      <td>[Protecting Privacy and Transforming COVID-19 ...</td>\n",
       "      <td>[We describe how CDC designed and produces 2 d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>33935584</td>\n",
       "      <td>8068562</td>\n",
       "      <td>A Novel Ensemble-based Classifier for Detectin...</td>\n",
       "      <td>The recently discovered coronavirus, SARS-CoV...</td>\n",
       "      <td>The SARS-CoV-2 that causes COVID-19, colloquia...</td>\n",
       "      <td>[A Novel Ensemble-based Classifier for Detecti...</td>\n",
       "      <td>[In this paper, a synthetic dataset of COVID-1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>33728416</td>\n",
       "      <td>7953447</td>\n",
       "      <td>VERSO: A comprehensive framework for the infer...</td>\n",
       "      <td>Summary  We introduce VERSO, a two-step frame...</td>\n",
       "      <td>The outbreak of coronavirus disease 2019 (COVI...</td>\n",
       "      <td>[VERSO: A comprehensive framework for the infe...</td>\n",
       "      <td>[The application to large-scale datasets demon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>35071990</td>\n",
       "      <td>8767797</td>\n",
       "      <td>Development and implementation of a COVID-19 c...</td>\n",
       "      <td>Introduction  Convalescent Plasma therapy is...</td>\n",
       "      <td>In late 2019, patients presenting an atypical ...</td>\n",
       "      <td>[Development and implementation of a COVID-19 ...</td>\n",
       "      <td>[Methods  A multicentric convalescent plasma c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>32898686</td>\n",
       "      <td>7474840</td>\n",
       "      <td>Digital technology, tele-medicine and artifici...</td>\n",
       "      <td>The simultaneous maturation of multiple digit...</td>\n",
       "      <td>2020 marked the synchronous maturation of seve...</td>\n",
       "      <td>[Digital technology, tele-medicine and artific...</td>\n",
       "      <td>[A review of different tele-health models and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>33302292</td>\n",
       "      <td>7799233</td>\n",
       "      <td>iDMer: an integrative and mechanism-driven res...</td>\n",
       "      <td>Abstract  Emerging viral infections seriously...</td>\n",
       "      <td>Emerging viral infections seriously threaten h...</td>\n",
       "      <td>[iDMer: an integrative and mechanism-driven re...</td>\n",
       "      <td>[As a one-stop integrative platform, iDMer inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>33052312</td>\n",
       "      <td>7543791</td>\n",
       "      <td>Virus database annotations assist in tracing i...</td>\n",
       "      <td>The global pandemic of SARS-CoV-2 has disrupt...</td>\n",
       "      <td>The coronavirus disease 2019 (COVID-19) pandem...</td>\n",
       "      <td>[Virus database annotations assist in tracing ...</td>\n",
       "      <td>[Virus database annotations assist in tracing ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>34228637</td>\n",
       "      <td>8312471</td>\n",
       "      <td>Prognostic and immunological value of  ATP6AP1...</td>\n",
       "      <td>Abnormal ATPase H+ Transporting Accessory Pro...</td>\n",
       "      <td>In women, breast cancer (BC) is one of the lea...</td>\n",
       "      <td>[Prognostic and immunological value of  ATP6AP...</td>\n",
       "      <td>[The Oncomine, Gene Expression Profiling Inter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>34451100</td>\n",
       "      <td>8402377</td>\n",
       "      <td>Explainable Artificial Intelligence for Bias D...</td>\n",
       "      <td>Problem: An application of Explainable Artifi...</td>\n",
       "      <td>The COVID-19 outbreak has become a central top...</td>\n",
       "      <td>[Explainable Artificial Intelligence for Bias ...</td>\n",
       "      <td>[Motivation: It is possible that classifiers a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>34335121</td>\n",
       "      <td>8317469</td>\n",
       "      <td>A stochastic bi-objective simulation–optimizat...</td>\n",
       "      <td>As of March 24, 2020, the Food and Drug Admin...</td>\n",
       "      <td>Nowadays we are challenging with the daily inc...</td>\n",
       "      <td>[A stochastic bi-objective simulation–optimiza...</td>\n",
       "      <td>[Therefore, a four-echelon supply chain has be...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pmid    pmcid                                              title  \\\n",
       "0   34847386  8609674  Uncertainty-aware convolutional neural network...   \n",
       "1   34976572  8675556  Detecting SARS-CoV-2 From Chest X-Ray Using Ar...   \n",
       "2   34352699  8249716  Conventional and microfluidic methods for airb...   \n",
       "3   32836689  7236753  Competitive pricing of substitute products und...   \n",
       "4   34070422  8226892  Shall I Work with Them? A Knowledge Graph-Base...   \n",
       "5   34139910  8216038  Protecting Privacy and Transforming COVID-19 C...   \n",
       "6   33935584  8068562  A Novel Ensemble-based Classifier for Detectin...   \n",
       "7   33728416  7953447  VERSO: A comprehensive framework for the infer...   \n",
       "8   35071990  8767797  Development and implementation of a COVID-19 c...   \n",
       "9   32898686  7474840  Digital technology, tele-medicine and artifici...   \n",
       "10  33302292  7799233  iDMer: an integrative and mechanism-driven res...   \n",
       "11  33052312  7543791  Virus database annotations assist in tracing i...   \n",
       "12  34228637  8312471  Prognostic and immunological value of  ATP6AP1...   \n",
       "13  34451100  8402377  Explainable Artificial Intelligence for Bias D...   \n",
       "14  34335121  8317469  A stochastic bi-objective simulation–optimizat...   \n",
       "\n",
       "                                             abstract  \\\n",
       "0    Deep learning (DL) has shown great success in...   \n",
       "1    Chest radiographs (X-rays) combined with Deep...   \n",
       "2    Graphical abstract      With the COVID-19 pan...   \n",
       "3    There has been an increased interest in optim...   \n",
       "4    We consider the prediction of future research...   \n",
       "5     Objectives  Federal open-data initiatives th...   \n",
       "6    The recently discovered coronavirus, SARS-CoV...   \n",
       "7    Summary  We introduce VERSO, a two-step frame...   \n",
       "8     Introduction  Convalescent Plasma therapy is...   \n",
       "9    The simultaneous maturation of multiple digit...   \n",
       "10   Abstract  Emerging viral infections seriously...   \n",
       "11   The global pandemic of SARS-CoV-2 has disrupt...   \n",
       "12   Abnormal ATPase H+ Transporting Accessory Pro...   \n",
       "13   Problem: An application of Explainable Artifi...   \n",
       "14   As of March 24, 2020, the Food and Drug Admin...   \n",
       "\n",
       "                                            main text  \\\n",
       "0   Novel Corona Virus disease 2019 (COVID-19) has...   \n",
       "1   The Severe Acute Respiratory Syndrome Coronavi...   \n",
       "2   Throughout history, infectious diseases have p...   \n",
       "3   Retailers frequently rely on suppliers that ar...   \n",
       "4   The development of knowledge graph-based appro...   \n",
       "5   Federal open-data initiatives that promote inc...   \n",
       "6   The SARS-CoV-2 that causes COVID-19, colloquia...   \n",
       "7   The outbreak of coronavirus disease 2019 (COVI...   \n",
       "8   In late 2019, patients presenting an atypical ...   \n",
       "9   2020 marked the synchronous maturation of seve...   \n",
       "10  Emerging viral infections seriously threaten h...   \n",
       "11  The coronavirus disease 2019 (COVID-19) pandem...   \n",
       "12  In women, breast cancer (BC) is one of the lea...   \n",
       "13  The COVID-19 outbreak has become a central top...   \n",
       "14  Nowadays we are challenging with the daily inc...   \n",
       "\n",
       "                                            sentences  \\\n",
       "0   [Uncertainty-aware convolutional neural networ...   \n",
       "1   [Detecting SARS-CoV-2 From Chest X-Ray Using A...   \n",
       "2   [Conventional and microfluidic methods for air...   \n",
       "3   [Competitive pricing of substitute products un...   \n",
       "4   [Shall I Work with Them? , A Knowledge Graph-B...   \n",
       "5   [Protecting Privacy and Transforming COVID-19 ...   \n",
       "6   [A Novel Ensemble-based Classifier for Detecti...   \n",
       "7   [VERSO: A comprehensive framework for the infe...   \n",
       "8   [Development and implementation of a COVID-19 ...   \n",
       "9   [Digital technology, tele-medicine and artific...   \n",
       "10  [iDMer: an integrative and mechanism-driven re...   \n",
       "11  [Virus database annotations assist in tracing ...   \n",
       "12  [Prognostic and immunological value of  ATP6AP...   \n",
       "13  [Explainable Artificial Intelligence for Bias ...   \n",
       "14  [A stochastic bi-objective simulation–optimiza...   \n",
       "\n",
       "                                          has_pattern  \n",
       "0   [The proposed method is evaluated on the three...  \n",
       "1   [However, questions remain regarding the accur...  \n",
       "2   [Creating targeted and effective interventions...  \n",
       "3   [Section 3 is devoted to benchmark modeling of...  \n",
       "4   [We benchmark the proposed approach against cl...  \n",
       "5   [We describe how CDC designed and produces 2 d...  \n",
       "6   [In this paper, a synthetic dataset of COVID-1...  \n",
       "7   [The application to large-scale datasets demon...  \n",
       "8   [Methods  A multicentric convalescent plasma c...  \n",
       "9   [A review of different tele-health models and ...  \n",
       "10  [As a one-stop integrative platform, iDMer inc...  \n",
       "11  [Virus database annotations assist in tracing ...  \n",
       "12  [The Oncomine, Gene Expression Profiling Inter...  \n",
       "13  [Motivation: It is possible that classifiers a...  \n",
       "14  [Therefore, a four-echelon supply chain has be...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "qualified_articles = []\n",
    "for article in articles:\n",
    "    if \"has_pattern\" in article and len(article['has_pattern']) > 20:\n",
    "        qualified_articles.append(article)\n",
    "\n",
    "qualified_df = pd.DataFrame(qualified_articles)\n",
    "display(qualified_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56163009",
   "metadata": {},
   "source": [
    "### Search dataset description sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ff0e870c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Searched 1000 full-text articles.\n",
      "Searched 2000 full-text articles.\n",
      "Searched 3000 full-text articles.\n",
      "Searched 4000 full-text articles.\n",
      "Searched 5000 full-text articles.\n",
      "Searched 6000 full-text articles.\n",
      "Searched 7000 full-text articles.\n",
      "Searched 8000 full-text articles.\n",
      "Searched 9000 full-text articles.\n",
      "Searched 10000 full-text articles.\n",
      "Searched 11000 full-text articles.\n",
      "Searched 12000 full-text articles.\n",
      "Searched 13000 full-text articles.\n",
      "Searched 14000 full-text articles.\n",
      "Searched 15000 full-text articles.\n",
      "Searched 16000 full-text articles.\n",
      "Searched 17000 full-text articles.\n",
      "Searched 18000 full-text articles.\n",
      "Searched 19000 full-text articles.\n",
      "Searched 20000 full-text articles.\n",
      "Searched 21000 full-text articles.\n",
      "Searched 22000 full-text articles.\n",
      "Searched 23000 full-text articles.\n",
      "Searched 24000 full-text articles.\n",
      "Searched 25000 full-text articles.\n",
      "Searched 26000 full-text articles.\n",
      "Searched 27000 full-text articles.\n",
      "Searched 28000 full-text articles.\n",
      "Searched 29000 full-text articles.\n",
      "Searched 30000 full-text articles.\n",
      "Searched 31000 full-text articles.\n",
      "Searched 32000 full-text articles.\n",
      "Searched 33000 full-text articles.\n",
      "Searched 34000 full-text articles.\n",
      "Searched 35000 full-text articles.\n",
      "Searched 36000 full-text articles.\n",
      "Searched 37000 full-text articles.\n",
      "Searched 38000 full-text articles.\n",
      "Searched 39000 full-text articles.\n",
      "Searched 40000 full-text articles.\n",
      "Searched 41000 full-text articles.\n",
      "Searched 42000 full-text articles.\n",
      "Searched 43000 full-text articles.\n",
      "Searched 44000 full-text articles.\n",
      "Searched 45000 full-text articles.\n",
      "Searched 46000 full-text articles.\n",
      "Searched 47000 full-text articles.\n",
      "Searched 48000 full-text articles.\n",
      "Searched 49000 full-text articles.\n",
      "Searched 50000 full-text articles.\n",
      "Searched 51000 full-text articles.\n",
      "Searched 52000 full-text articles.\n",
      "Searched 53000 full-text articles.\n",
      "Searched 54000 full-text articles.\n",
      "Searched 55000 full-text articles.\n",
      "Searched 56000 full-text articles.\n",
      "Searched 57000 full-text articles.\n",
      "Searched 58000 full-text articles.\n",
      "Searched 59000 full-text articles.\n",
      "Searched 60000 full-text articles.\n",
      "Searched 61000 full-text articles.\n",
      "Search Completed.\n"
     ]
    }
   ],
   "source": [
    "seg = pysbd.Segmenter(language=\"en\", clean=False)\n",
    "pattern = re.compile(r'(N3C)|(National COVID Cohort Collaborative)|(GenBank)|\\\n",
    "                     (GISAID)|(Nextstrain)|(OpenICPSR)|(TCIA)|(GEO)|(Gene Expression Omnibus)\\\n",
    "                     (CORD-19)|(LitCovid)|(NCBI Virus)')\n",
    "\n",
    "for i, article in enumerate(articles):\n",
    "    # all_text = article['title'] + \".\\n\" + article['abstract'] + \"\\n\" + article['main text']\n",
    "    title_abstract_text = article['title'] + \".\\n\" + article['abstract']\n",
    "    article['sentences'] = seg.segment(title_abstract_text)\n",
    "    # article['sentences'] = seg.segment(all_text)\n",
    "    article['has_pattern'] = []\n",
    "    for j, sent in enumerate(article['sentences']):\n",
    "        if pattern.search(sent):\n",
    "            article['has_pattern'].append(sent)\n",
    "    if i != 0 and i % 1000 == 0:\n",
    "        print(\"Searched %d full-text articles.\" % i)\n",
    "        # break\n",
    "\n",
    "print(\"Search Completed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "032a5511",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>main text</th>\n",
       "      <th>sentences</th>\n",
       "      <th>has_pattern</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34765026</td>\n",
       "      <td>8576622</td>\n",
       "      <td>Testicular injury during SARS-CoV-2 infection ...</td>\n",
       "      <td>Coronavirus disease 2019 (COVID-19), caused b...</td>\n",
       "      <td>Coronavirus disease 2019 (COVID-19), caused by...</td>\n",
       "      <td>[Testicular injury during SARS-CoV-2 infection...</td>\n",
       "      <td>[For this, the ACE2-expressing cell compositio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>35422859</td>\n",
       "      <td>9002903</td>\n",
       "      <td>Combination of Enrichment Using Gene Ontology ...</td>\n",
       "      <td>Introduction  The severity of coronavirus di...</td>\n",
       "      <td>A coronavirus is a group of viruses from the s...</td>\n",
       "      <td>[Combination of Enrichment Using Gene Ontology...</td>\n",
       "      <td>[We collected the transcriptomics data from GE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33166392</td>\n",
       "      <td>7778958</td>\n",
       "      <td>LitCovid: an open database of COVID-19 literature</td>\n",
       "      <td>Abstract  Since the outbreak of the current p...</td>\n",
       "      <td>Since the outbreak of the COVID-19 pandemic, r...</td>\n",
       "      <td>[LitCovid: an open database of COVID-19 litera...</td>\n",
       "      <td>[LitCovid: an open database of COVID-19 litera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>34821594</td>\n",
       "      <td>8625834</td>\n",
       "      <td>Characterization and Pathogenic Speculation of...</td>\n",
       "      <td>Patients with coronavirus disease 2019 (COVID...</td>\n",
       "      <td>Infection with severe acute respiratory syndro...</td>\n",
       "      <td>[Characterization and Pathogenic Speculation o...</td>\n",
       "      <td>[Scientific articles were retrieved by searchi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32022275</td>\n",
       "      <td>7166327</td>\n",
       "      <td>The first two cases of 2019‐nCoV in Italy: Whe...</td>\n",
       "      <td>Abstract  A novel Coronavirus , 2019‐nCoV, ha...</td>\n",
       "      <td>An ongoing epidemic by a new Coronavirus, name...</td>\n",
       "      <td>[The first two cases of 2019‐nCoV in Italy: Wh...</td>\n",
       "      <td>[A maximum clade credibility tree has been bui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>35005058</td>\n",
       "      <td>8734487</td>\n",
       "      <td>Evaluation of multiple open-source deep learni...</td>\n",
       "      <td>Abstract.  Purpose : Chest x-rays are complex...</td>\n",
       "      <td>The early radiological features of viral pneum...</td>\n",
       "      <td>[Evaluation of multiple open-source deep learn...</td>\n",
       "      <td>[Approach : We tested three open-source deep l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>34675627</td>\n",
       "      <td>8520483</td>\n",
       "      <td>Construction and Investigation of Competing En...</td>\n",
       "      <td>Introduction  The current COVID-19 pandemic ...</td>\n",
       "      <td>The Coronaviruses (CoVs) are a varied group of...</td>\n",
       "      <td>[Construction and Investigation of Competing E...</td>\n",
       "      <td>[Methods  The RNA sequencing data of SARS-CoV-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>32773643</td>\n",
       "      <td>7493783</td>\n",
       "      <td>Genomic variance of Open Reading Frames (ORFs)...</td>\n",
       "      <td>Background:  The outbreak of severe acute res...</td>\n",
       "      <td>The outbreak of severe acute respiratory syndr...</td>\n",
       "      <td>[Genomic variance of Open Reading Frames (ORFs...</td>\n",
       "      <td>[Methods:  We apply genomic alignment analysis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>34019643</td>\n",
       "      <td>8218201</td>\n",
       "      <td>eVITTA: a web-based visualization and inferenc...</td>\n",
       "      <td>Abstract  Transcriptome profiling is essentia...</td>\n",
       "      <td>Transcriptome profiling is an essential techni...</td>\n",
       "      <td>[eVITTA: a web-based visualization and inferen...</td>\n",
       "      <td>[Additionally, there is no systematic way to e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>32946531</td>\n",
       "      <td>7500631</td>\n",
       "      <td>The association between ABO blood group and SA...</td>\n",
       "      <td>At present, existing evidence about the assoc...</td>\n",
       "      <td>Since the first outbreak of COVID-19 in Wuhan,...</td>\n",
       "      <td>[The association between ABO blood group and S...</td>\n",
       "      <td>[We performed a systematic search on MEDLINE a...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pmid    pmcid                                              title  \\\n",
       "0   34765026  8576622  Testicular injury during SARS-CoV-2 infection ...   \n",
       "1   35422859  9002903  Combination of Enrichment Using Gene Ontology ...   \n",
       "2   33166392  7778958  LitCovid: an open database of COVID-19 literature   \n",
       "3   34821594  8625834  Characterization and Pathogenic Speculation of...   \n",
       "4   32022275  7166327  The first two cases of 2019‐nCoV in Italy: Whe...   \n",
       "..       ...      ...                                                ...   \n",
       "69  35005058  8734487  Evaluation of multiple open-source deep learni...   \n",
       "70  34675627  8520483  Construction and Investigation of Competing En...   \n",
       "71  32773643  7493783  Genomic variance of Open Reading Frames (ORFs)...   \n",
       "72  34019643  8218201  eVITTA: a web-based visualization and inferenc...   \n",
       "73  32946531  7500631  The association between ABO blood group and SA...   \n",
       "\n",
       "                                             abstract  \\\n",
       "0    Coronavirus disease 2019 (COVID-19), caused b...   \n",
       "1     Introduction  The severity of coronavirus di...   \n",
       "2    Abstract  Since the outbreak of the current p...   \n",
       "3    Patients with coronavirus disease 2019 (COVID...   \n",
       "4    Abstract  A novel Coronavirus , 2019‐nCoV, ha...   \n",
       "..                                                ...   \n",
       "69   Abstract.  Purpose : Chest x-rays are complex...   \n",
       "70    Introduction  The current COVID-19 pandemic ...   \n",
       "71   Background:  The outbreak of severe acute res...   \n",
       "72   Abstract  Transcriptome profiling is essentia...   \n",
       "73   At present, existing evidence about the assoc...   \n",
       "\n",
       "                                            main text  \\\n",
       "0   Coronavirus disease 2019 (COVID-19), caused by...   \n",
       "1   A coronavirus is a group of viruses from the s...   \n",
       "2   Since the outbreak of the COVID-19 pandemic, r...   \n",
       "3   Infection with severe acute respiratory syndro...   \n",
       "4   An ongoing epidemic by a new Coronavirus, name...   \n",
       "..                                                ...   \n",
       "69  The early radiological features of viral pneum...   \n",
       "70  The Coronaviruses (CoVs) are a varied group of...   \n",
       "71  The outbreak of severe acute respiratory syndr...   \n",
       "72  Transcriptome profiling is an essential techni...   \n",
       "73  Since the first outbreak of COVID-19 in Wuhan,...   \n",
       "\n",
       "                                            sentences  \\\n",
       "0   [Testicular injury during SARS-CoV-2 infection...   \n",
       "1   [Combination of Enrichment Using Gene Ontology...   \n",
       "2   [LitCovid: an open database of COVID-19 litera...   \n",
       "3   [Characterization and Pathogenic Speculation o...   \n",
       "4   [The first two cases of 2019‐nCoV in Italy: Wh...   \n",
       "..                                                ...   \n",
       "69  [Evaluation of multiple open-source deep learn...   \n",
       "70  [Construction and Investigation of Competing E...   \n",
       "71  [Genomic variance of Open Reading Frames (ORFs...   \n",
       "72  [eVITTA: a web-based visualization and inferen...   \n",
       "73  [The association between ABO blood group and S...   \n",
       "\n",
       "                                          has_pattern  \n",
       "0   [For this, the ACE2-expressing cell compositio...  \n",
       "1   [We collected the transcriptomics data from GE...  \n",
       "2   [LitCovid: an open database of COVID-19 litera...  \n",
       "3   [Scientific articles were retrieved by searchi...  \n",
       "4   [A maximum clade credibility tree has been bui...  \n",
       "..                                                ...  \n",
       "69  [Approach : We tested three open-source deep l...  \n",
       "70  [Methods  The RNA sequencing data of SARS-CoV-...  \n",
       "71  [Methods:  We apply genomic alignment analysis...  \n",
       "72  [Additionally, there is no systematic way to e...  \n",
       "73  [We performed a systematic search on MEDLINE a...  \n",
       "\n",
       "[74 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "qualified_articles = []\n",
    "for article in articles:\n",
    "    if \"has_pattern\" in article and len(article['has_pattern']) > 0:\n",
    "        qualified_articles.append(article)\n",
    "\n",
    "qualified_df = pd.DataFrame(qualified_articles)\n",
    "display(qualified_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6954fd87",
   "metadata": {},
   "outputs": [],
   "source": [
    "txt_directory = \"./fulltext_txt_nih/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ac3ed10",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "string indices must be integers",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_78585/2842238824.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0marticle\u001b[0m \u001b[0;32min\u001b[0m \u001b[0marticle_batch\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;31m# print(article['pmcid'])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0mall_text\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marticle\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'title'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\".\\n\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0marticle\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'abstract'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"\\n\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0marticle\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'main text'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0;31m# Format lines.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mraw_lines\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mall_text\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplitlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: string indices must be integers"
     ]
    }
   ],
   "source": [
    "# parser = CoreNLPParser(url='http://localhost:9000')\n",
    "\n",
    "for article in article_batch:\n",
    "    # print(article['pmcid'])\n",
    "    all_text = article['title'] + \".\\n\" + article['abstract'] + \"\\n\" + article['main text'].strip()\n",
    "    # Format lines.\n",
    "    raw_lines = all_text.splitlines()\n",
    "    # lines = [line.strip() for line in raw_lines if len(word_tokenize(line)) > 10]\n",
    "    lines = []\n",
    "    for line in raw_lines:\n",
    "        line_decode = unidecode.unidecode(line)\n",
    "        if len(word_tokenize(line_decode)) > 10:\n",
    "            lines.append(line.strip())\n",
    "        else:\n",
    "            continue\n",
    "    new_lines = []\n",
    "    for line in lines:\n",
    "        if not new_lines or not(line.startswith((\")\", \"-\", \",\"))):\n",
    "            new_lines.append(line)\n",
    "        else:\n",
    "            new_lines[-1] += line\n",
    "    new_text_per_article = \"\\n\\n\".join(new_lines)\n",
    "    txt_name = \"PMC{}.txt\".format(article['pmcid'])\n",
    "    output_path = os.path.join(txt_directory, txt_name)\n",
    "    with open(output_path, \"w\", encoding='utf-8') as output_file:\n",
    "        output_file.write(new_text_per_article)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26696693",
   "metadata": {},
   "source": [
    "## Detailed search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1bb4d238",
   "metadata": {},
   "outputs": [],
   "source": [
    "from nltk.tokenize import sent_tokenize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "159bfba6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PMC8675677.xml', 'PMC7989622.xml', 'PMC8180385.xml', 'PMC8336924.xml', 'PMC8653328.xml', 'PMC8362655.xml', 'PMC8216038.xml', 'PMC8319175.xml', 'PMC7823180.xml', 'PMC8468495.xml']\n"
     ]
    }
   ],
   "source": [
    "pmcs_has_pattern = top20_df['pmcid'].to_list()\n",
    "pmcpath_has_pattern = [\"PMC\" + pmcid + \".xml\" for pmcid in pmcs_has_pattern]\n",
    "print(pmcpath_has_pattern[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "66528997",
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0\n",
    "count_all = 0\n",
    "dataset_sents = []\n",
    "\n",
    "for xml_name in pmcpath_has_pattern[:10]:\n",
    "    xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    metadata = pp.parse_pubmed_xml(xml_path)\n",
    "    full_text = pp.parse_pubmed_paragraph(xml_path, all_paragraph=True)\n",
    "    for para in full_text:\n",
    "        sents = sent_tokenize(para['text'])\n",
    "        for sent in sents:\n",
    "            count_all += 1\n",
    "            if re.search(pattern, sent):\n",
    "                count += 1\n",
    "                sentence = {\"pmid\": metadata['pmid'], \"pmcid\": metadata['pmc'], \"sentence\": sent}\n",
    "                dataset_sents.append(sentence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "228bad23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>sentence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34914688</td>\n",
       "      <td>8675677</td>\n",
       "      <td>Using this graphical model called the Stochast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>34914688</td>\n",
       "      <td>8675677</td>\n",
       "      <td>To this end, we first analyzed a large dataset...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34914688</td>\n",
       "      <td>8675677</td>\n",
       "      <td>We then analyzed other datasets to determine w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>34914688</td>\n",
       "      <td>8675677</td>\n",
       "      <td>Early in the pandemic, we implemented the Stoc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>34914688</td>\n",
       "      <td>8675677</td>\n",
       "      <td>We chose to first perform the same analysis ov...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>851</th>\n",
       "      <td>34573790</td>\n",
       "      <td>8468495</td>\n",
       "      <td>For our analysis, we used the synthetic and re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>852</th>\n",
       "      <td>34573790</td>\n",
       "      <td>8468495</td>\n",
       "      <td>Additional synthetic and real datasets exist b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>34573790</td>\n",
       "      <td>8468495</td>\n",
       "      <td>The synthetic datasets for the three different...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>34573790</td>\n",
       "      <td>8468495</td>\n",
       "      <td>The synthetic versions of some datasets and th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>855</th>\n",
       "      <td>34573790</td>\n",
       "      <td>8468495</td>\n",
       "      <td>The real MIMIC and ASD datasets are only acces...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>856 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pmid    pmcid                                           sentence\n",
       "0    34914688  8675677  Using this graphical model called the Stochast...\n",
       "1    34914688  8675677  To this end, we first analyzed a large dataset...\n",
       "2    34914688  8675677  We then analyzed other datasets to determine w...\n",
       "3    34914688  8675677  Early in the pandemic, we implemented the Stoc...\n",
       "4    34914688  8675677  We chose to first perform the same analysis ov...\n",
       "..        ...      ...                                                ...\n",
       "851  34573790  8468495  For our analysis, we used the synthetic and re...\n",
       "852  34573790  8468495  Additional synthetic and real datasets exist b...\n",
       "853  34573790  8468495  The synthetic datasets for the three different...\n",
       "854  34573790  8468495  The synthetic versions of some datasets and th...\n",
       "855  34573790  8468495  The real MIMIC and ASD datasets are only acces...\n",
       "\n",
       "[856 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sent_df = pd.DataFrame(dataset_sents)\n",
    "display(sent_df)\n",
    "sent_df.to_csv('./top10_sents.tsv', sep='\\t', index=False)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5acc143e",
   "metadata": {},
   "source": [
    "## Parse full-text with section headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8cc8d5e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PMC8475511.xml', 'PMC8609674.xml', 'PMC8675556.xml', 'PMC7236753.xml', 'PMC8216038.xml', 'PMC8767797.xml', 'PMC7474840.xml', 'PMC7799233.xml', 'PMC7543791.xml', 'PMC8312471.xml']\n"
     ]
    }
   ],
   "source": [
    "# Generate PMC article file names.\n",
    "\n",
    "# pmc_to_extract = df_rand_100['pmcid'].to_list()\n",
    "pmc_to_extract = remain_articles\n",
    "pmc_to_extract_filenames = [\"PMC\" + pmcid + \".xml\" for pmcid in pmc_to_extract]\n",
    "print(pmc_to_extract_filenames[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ede0e591",
   "metadata": {},
   "outputs": [],
   "source": [
    "txt_directory = \"./fulltext_txt/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0c71e107",
   "metadata": {},
   "outputs": [],
   "source": [
    "article_batch = [\"PMC8491001.xml\", \"PMC9169748.xml\", \"PMC7970480.xml\", \"PMC8662089.xml\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b18d2311",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate text files for annotation.\n",
    "\n",
    "for i, xml_name in enumerate(article_batch):\n",
    "    xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    metadata = pp.parse_pubmed_xml(xml_path)\n",
    "    full_text = pp.parse_pubmed_paragraph(xml_path, all_paragraph=True)\n",
    "    \n",
    "    paras = []\n",
    "    for para in full_text:\n",
    "        paras.append(para['text'].strip())\n",
    "        main_text = \"\\n\".join(paras)\n",
    "        # paras.append(paragraph['all text'])\n",
    "    all_text_per_article = metadata['full_title'] + \".\\n\" + metadata['abstract'] + \"\\n\" + main_text\n",
    "    \n",
    "    # Format lines.\n",
    "    lines = all_text_per_article.splitlines()\n",
    "    lines = [line for line in lines if len(line) > 2]\n",
    "    new_lines = []\n",
    "    for line in lines:\n",
    "        if not new_lines or not(line.startswith((\")\", \"-\", \",\"))):\n",
    "            new_lines.append(line)\n",
    "        else:\n",
    "            new_lines[-1] += line\n",
    "    new_text_per_article = \"\\n\\n\".join(new_lines)\n",
    "    # print(all_text_per_article)\n",
    "    \n",
    "    txt_name = xml_name.replace(\".xml\", \".txt\")\n",
    "    output_path = os.path.join(txt_directory, txt_name)\n",
    "    with open(output_path, \"w\", encoding='utf-8') as output_file:\n",
    "        output_file.write(new_text_per_article)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb187f2e",
   "metadata": {},
   "source": [
    "## Parse out Mesh and Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1355257d",
   "metadata": {},
   "outputs": [],
   "source": [
    "section_pattern = re.compile(r\"\\bmethods?\\b|\\bMethodolog[yi]\\b\", re.IGNORECASE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f52b9adb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rwd_sec_query = re.compile(r'\\b(CMS|Medicare|Medicaid|SEER|Optum|Cinformatics|Data Mart|MIMIC|NHANES|National Health Survey|Nutrition Examination Survey|NAMCS|Ambulatory Medical Care Survey|NIS|Inpatient Sample|VHA|CDW|Veterans Health Administration|OptumLabs|Data Warehouse|N3C|National COVID Cohort Collaborative)\\b', re.IGNORECASE)\n",
    "# omics_sec_query = re.compile(r'(sequence read archive|SRA)|covid-19 host genetics initiative|covid-19 cell atlas|(european nucleotide archive|ENA)|broad terra|2019nCoVR|(covid-19 genome sequence dataset)|(european bioinformatics institute|embl-ebi)|genbank|gaid|gisa(\\\\w+)|ncbi virus|nextstrain|reactome')\n",
    "rwd_sec_query = re.compile(r'\\b(Optum|Cinformatics|Data Mart|MIMIC|National Health Survey|Nutrition Examination Survey|Ambulatory Medical Care Survey|Corporate Data Warehouse|N3C|National COVID Cohort Collaborative)\\b')\n",
    "omics_sec_query = re.compile(r'(Sequence Read Archive|SRA)|(Host Genetics Initiative|HGI)|COVID-19 Cell Atlas|(European Nucleotide Archive|ENA)|broad terra|2019nCoVR|(COVID-19 Data Portal)|[Gg]en[Bb]ank|(GISAID)|(Gene Expression Omnibus|GEO)|(\\b[Nn]extstrain\\b)|Reactome|(NCBI Virus)|(PubChem)|(Kyoto Encyclopedia of Genes and Genomes|KEGG))')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9fe43a81",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'find'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_56294/2272687765.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mpmc_tree\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0metree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpmc_xml_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mmeta\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpmc_tree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\".//article-meta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0mmetadata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_pubmed_xml\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpmc_xml_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m     \u001b[0mfull_text\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_pubmed_paragraph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpmc_xml_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mall_paragraph\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mparas\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages/pubmed_parser/pubmed_oa_parser.py\u001b[0m in \u001b[0;36mparse_pubmed_xml\u001b[0;34m(path, include_path, nxml)\u001b[0m\n\u001b[1;32m    155\u001b[0m         \u001b[0mjournal\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    156\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 157\u001b[0;31m     \u001b[0mdict_article_meta\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_article_meta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtree\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    158\u001b[0m     \u001b[0mpub_year_node\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\".//pub-date/year\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    159\u001b[0m     \u001b[0mpub_year\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpub_year_node\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mpub_year_node\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages/pubmed_parser/pubmed_oa_parser.py\u001b[0m in \u001b[0;36mparse_article_meta\u001b[0;34m(tree)\u001b[0m\n\u001b[1;32m     67\u001b[0m     \"\"\"\n\u001b[1;32m     68\u001b[0m     \u001b[0marticle_meta\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\".//article-meta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 69\u001b[0;31m     \u001b[0mpmid_node\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marticle_meta\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'article-id[@pub-id-type=\"pmid\"]'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     70\u001b[0m     \u001b[0mpmc_node\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marticle_meta\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'article-id[@pub-id-type=\"pmc\"]'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m     \u001b[0mpub_id_node\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marticle_meta\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'article-id[@pub-id-type=\"publisher-id\"]'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'find'"
     ]
    }
   ],
   "source": [
    "# Parse out text.\n",
    "articles = []\n",
    "\n",
    "for i, xml_name in enumerate(current_filelist):\n",
    "    pmc_xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    pmc_tree = etree.parse(pmc_xml_path)\n",
    "    meta = pmc_tree.find(\".//article-meta\")\n",
    "    metadata = pp.parse_pubmed_xml(pmc_xml_path)\n",
    "    full_text = pp.parse_pubmed_paragraph(pmc_xml_path, all_paragraph=True)\n",
    "    paras = []\n",
    "    for j, para in enumerate(full_text):\n",
    "        if section_pattern.search(para['section']):\n",
    "            paras.append(para['text'])\n",
    "        else:\n",
    "            continue\n",
    "    main_text = \"\\n\".join(paras)\n",
    "    pmc_article = {\"pmid\": metadata['pmid'], \"pmcid\": metadata['pmc'], \n",
    "                       \"title\": metadata['full_title'], \"abstract\": metadata['abstract'], \n",
    "                       \"main text\": main_text}\n",
    "    articles.append(pmc_article)\n",
    "    \n",
    "\n",
    "print(\"Found %d articles.\" % len(articles))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3dcbfb3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "140649\n"
     ]
    }
   ],
   "source": [
    "print(len(articles))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5b019462",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8014456\tSRA\n",
      "8673727\tGISAID\n",
      "8598683\tMedicare\n",
      "7853258\tGISAID\n",
      "8847733\tGISAID\n",
      "8222831\tGenbank\n",
      "7405216\tNutrition Examination Survey\n",
      "8156524\tReactome\n",
      "8491001\tVHA\n",
      "9169748\tGEO\n",
      "7970480\tData Warehouse\n",
      "8662089\tENA\n",
      "7898962\tMedicare\n",
      "9088623\tnextstrain\n",
      "8689746\tGISAID\n",
      "7454936\tGISAID\n",
      "8520125\tMedicaid\n",
      "8235306\tnextstrain\n",
      "8876352\tGISAID\n",
      "8351078\tData Warehouse\n",
      "7662030\tGEO\n",
      "8168855\tHost Genetics Initiative\n",
      "8065152\tGISAID\n",
      "8812349\tMedicaid\n",
      "7953981\tGISAID\n",
      "8420260\tENA\n",
      "8096765\tGISAID\n",
      "8014828\tGISAID\n",
      "8795823\tVHA\n",
      "7675408\tGISAID\n",
      "7099638\tGISAID\n",
      "8600389\tMedicare\n",
      "8755559\tGISAID\n",
      "7450963\tGISAID\n",
      "8080116\tMedicare\n",
      "8250548\tSRA\n",
      "9021587\tVHA\n",
      "8522425\tGISAID\n",
      "8920877\tGISAID\n",
      "7745650\tGISAID\n",
      "7404845\tMedicare\n",
      "8299148\tMedicare\n",
      "8307577\tGISAID\n",
      "7653226\tData Warehouse\n",
      "8014766\tMedicare\n",
      "8341697\tGISAID\n",
      "8919776\tMedicaid\n",
      "7825889\tGene Expression Omnibus\n",
      "8830142\tMedicare\n",
      "7716286\tMedicaid\n",
      "7851715\tGISAID\n",
      "9056975\tGISAID\n",
      "8601222\tGISAID\n",
      "7313917\tGISAID\n",
      "7861616\tnextstrain\n",
      "8649449\tMedicare\n",
      "7335245\tSRA\n",
      "8888350\tNextstrain\n",
      "7816100\tData Warehouse\n",
      "8573839\tGISAID\n",
      "8556844\tGISAID\n",
      "7936576\tMedicare\n",
      "7722480\tData Warehouse\n",
      "8806999\tGISAID\n",
      "8402964\tENA\n",
      "8744562\tENA\n",
      "8517740\tMedicare\n",
      "8977504\tCMS\n",
      "7945608\tCMS\n",
      "9012983\tNational COVID Cohort Collaborative\n",
      "7375210\tGISAID\n",
      "8041018\tMedicare\n",
      "8388146\tGISAID\n",
      "9013836\tGISAID\n",
      "8813191\tGISAID\n",
      "8353976\tData Warehouse\n",
      "7586862\tNutrition Examination Survey\n",
      "8939770\tGISAID\n",
      "7912354\tSRA\n",
      "7572353\tGISAID\n",
      "8287406\tCMS\n",
      "7857074\tReactome\n",
      "8620562\tGISAID\n",
      "8752051\tSRA\n",
      "8065384\tCMS\n",
      "8214951\tGISAID\n",
      "8047389\tCMS\n",
      "7857998\tNIS\n",
      "8793136\tGene Expression Omnibus\n",
      "9115045\tMedicare\n",
      "8278821\tGene Expression Omnibus\n",
      "8742307\tVeterans Health Administration\n",
      "8054526\tGene Expression Omnibus\n",
      "7812423\tENA\n",
      "9055753\tMIMIC\n",
      "9025597\tMedicare\n",
      "8689887\tCMS\n",
      "8824255\tMedicaid\n",
      "8427377\tData Mart\n",
      "8699860\tData Warehouse\n",
      "7833077\tNIS\n",
      "7831623\tMedicare\n",
      "8949797\tGISAID\n",
      "8547589\tMedicare\n",
      "7489837\tNHANES\n",
      "7573595\tReactome\n",
      "7717307\tGenbank\n",
      "8410838\tReactome\n",
      "7536974\tENA\n",
      "9014086\tGISAID\n",
      "8936035\tMedicare\n",
      "7472216\tGISAID\n",
      "8013415\tNutrition Examination Survey\n",
      "8795775\tGISAID\n",
      "8971476\tGISAID\n",
      "8881767\tNextstrain\n",
      "8323550\tMedicare\n",
      "8500154\tGISAID\n",
      "8558427\tData Warehouse\n",
      "7445146\tNIS\n",
      "8467961\tData Warehouse\n",
      "7454165\tNational Health Survey\n",
      "7794913\tGene Expression Omnibus\n",
      "8011323\tGISAID\n",
      "8638998\tGISAID\n",
      "8310374\tGISAID\n",
      "7300971\tGISAID\n",
      "7808963\tCMS\n",
      "7604091\tENA\n",
      "8594401\tMedicare\n",
      "8495534\tOptumLabs\n",
      "7816961\tMedicare\n",
      "8000869\tNational Health Survey\n",
      "7436865\tGene Expression Omnibus\n",
      "8394820\tnextstrain\n",
      "7885630\tCMS\n",
      "8629594\tCMS\n",
      "8378281\tMedicare\n",
      "7905771\tNHANES\n",
      "8916698\tData Warehouse\n",
      "8589371\tGISAID\n",
      "8899842\tGEO\n",
      "8564791\tENA\n",
      "8286246\tENA\n",
      "7527908\tNutrition Examination Survey\n",
      "8192838\tGISAID\n",
      "8847881\tVeterans Health Administration\n",
      "8955736\tGISAID\n",
      "8783435\tGISAID\n",
      "7900618\tGISAID\n",
      "9111858\tMedicaid\n",
      "8769737\tCMS\n",
      "8790551\tENA\n",
      "8038918\tMedicare\n",
      "8890914\tCDW\n",
      "8185546\tGISAID\n",
      "8517744\tData Warehouse\n",
      "7836769\tData Warehouse\n",
      "8678258\tData Mart\n",
      "8635622\tGenbank\n",
      "8155195\tMedicare\n",
      "8425019\tMedicare\n",
      "7232926\tGISAID\n",
      "8665655\tMedicare\n",
      "8009592\tGISAID\n",
      "8386796\tGISAID\n",
      "8440941\tSRA\n",
      "7384633\tData Warehouse\n",
      "8993036\tGISAID\n",
      "8008778\tData Mart\n",
      "7434426\tVHA\n",
      "7424139\tGEO\n",
      "8160489\tVHA\n",
      "7463225\tGISAID\n",
      "7387429\tGISAID\n",
      "7568500\tMedicaid\n",
      "8251179\tMedicare\n",
      "8567110\tVHA\n",
      "8472500\tENA\n",
      "8312728\tData Warehouse\n",
      "8251009\tMedicaid\n",
      "7985685\tGISAID\n",
      "7588421\tGISAID\n",
      "8321215\tVHA\n",
      "8283561\tVHA\n",
      "8845007\tMedicare\n",
      "8233550\tGene Expression Omnibus\n",
      "8482831\tENA\n",
      "7713534\tMedicare\n",
      "7112083\tGISAID\n",
      "8930830\tGISAID\n",
      "8914125\tENA\n",
      "8590495\tCMS\n",
      "7537892\tMedicaid\n",
      "7454314\tReactome\n",
      "9030035\tGISAID\n",
      "8128092\tGISAID\n",
      "7314513\tGene Expression Omnibus\n",
      "8743084\tMedicaid\n",
      "7766317\tMedicaid\n",
      "8409597\tVeterans Health Administration\n",
      "8483812\tGEO\n",
      "9021586\tData Warehouse\n",
      "8462492\tData Warehouse\n",
      "9043628\tMedicare\n",
      "7242205\tGene Expression Omnibus\n",
      "8734054\tGISAID\n",
      "7794290\tGEO\n",
      "8841901\tMedicaid\n",
      "8373619\tENA\n",
      "8673923\tNextstrain\n",
      "8447078\tMedicare\n",
      "7301886\tGene Expression Omnibus\n",
      "8989256\tMedicare\n",
      "9757901\tMedicare\n",
      "8016908\tGISAID\n",
      "7842178\tGene Expression Omnibus\n",
      "7914631\tGISAID\n",
      "7267315\tGISAID\n",
      "8143446\tGISAID\n",
      "7931211\tMedicare\n",
      "8660178\tMedicare\n",
      "8846540\tMedicare\n",
      "9035897\tGene Expression Omnibus\n",
      "8107815\tMedicaid\n",
      "7365377\tGene Expression Omnibus\n",
      "8386106\tENA\n",
      "8454282\tMedicare\n",
      "7386276\tGISAID\n",
      "8638233\tMedicare\n",
      "8544632\tMedicaid\n",
      "8907386\tGISAID\n",
      "8895335\tVeterans Health Administration\n",
      "7837224\tMedicare\n",
      "8441837\tVHA\n",
      "8702432\tMedicaid\n",
      "8802155\tENA\n",
      "9023376\tVeterans Health Administration\n",
      "7227581\tGISAID\n",
      "7232955\tGISAID\n",
      "8565135\tCMS\n",
      "7901655\tMedicare\n",
      "8206341\tGISAID\n",
      "8410951\tSRA\n",
      "7890668\tN3C\n",
      "8706409\tGene Expression Omnibus\n",
      "8883310\tN3C\n",
      "8937622\tVHA\n",
      "8655495\tGISAID\n",
      "7462517\tGISAID\n",
      "7899228\tGISAID\n",
      "8330918\tMedicare\n",
      "7400710\tGISAID\n",
      "8883040\tVeterans Health Administration\n",
      "7497891\tGEO\n",
      "8447318\tMedicare\n",
      "7860999\tMedicaid\n",
      "8416297\tGISAID\n",
      "8314823\tMedicare\n",
      "8486159\tMedicare\n",
      "7330599\tGEO\n",
      "7816561\tVHA\n",
      "8701661\tVHA\n",
      "7828593\tCMS\n",
      "9007721\tENA\n",
      "7895695\tGISAID\n",
      "8755353\tGISAID\n",
      "8440190\tGISAID\n",
      "9031584\tGISAID\n",
      "8474700\tENA\n",
      "8996281\tN3C\n",
      "8735795\tN3C\n",
      "7902226\tCMS\n",
      "8325377\tMedicaid\n",
      "7967077\tMedicare\n",
      "8837475\tnextstrain\n",
      "7853753\tVHA\n",
      "7693378\tGISAID\n",
      "8763401\tGEO\n",
      "9542480\tGISAID\n",
      "8763408\tMedicare\n",
      "9728395\tGISAID\n",
      "8845004\tGISAID\n",
      "8877658\tGISAID\n",
      "9115369\tGISAID\n",
      "8297137\tMedicare\n",
      "8065242\tGISAID\n",
      "8908351\tGISAID\n",
      "8542493\tVHA\n",
      "7808873\tGISAID\n",
      "8537103\tGISAID\n",
      "7286220\tMedicaid\n",
      "8314815\tnextstrain\n",
      "8532098\tSRA\n",
      "8616735\tData Warehouse\n",
      "7989431\tMedicare\n",
      "8720299\tENA\n",
      "8740823\tSRA\n",
      "7530441\tGISAID\n",
      "7954359\tSRA\n",
      "8761839\tVeterans Health Administration\n",
      "8354098\tGISAID\n",
      "7462411\tGene Expression Omnibus\n",
      "7195303\tGISAID\n",
      "8370846\tN3C\n",
      "8972875\tGISAID\n",
      "8976579\tNIS\n",
      "8145975\tGISAID\n",
      "7106073\tGISAID\n",
      "8011776\tVeterans Health Administration\n",
      "7564308\tCMS\n",
      "8357963\tReactome\n",
      "7499577\tData Warehouse\n",
      "8739812\tMedicaid\n",
      "7332916\tMedicare\n",
      "8463310\tGene Expression Omnibus\n",
      "8008803\tGene Expression Omnibus\n",
      "7537300\tGISAID\n",
      "8888211\tGISAID\n",
      "8950183\tGISAID\n",
      "7999510\tGISAID\n",
      "8470798\tENA\n",
      "8968467\tGISAID\n",
      "8617653\tNational Health Survey\n",
      "8329552\tnextstrain\n",
      "8880534\tnextstrain\n",
      "8017083\tGISAID\n",
      "9257501\tMedicaid\n",
      "8917452\tGISAID\n",
      "8653020\tN3C\n",
      "8276543\tENA\n",
      "7194697\tCMS\n",
      "8267241\tGISAID\n",
      "8233406\tMedicare\n",
      "8244812\tGISAID\n",
      "9088542\tGISAID\n",
      "8023754\tMedicaid\n",
      "8515932\tGISAID\n",
      "7429414\tMedicaid\n",
      "9760484\tCMS\n",
      "8542497\tMedicare\n",
      "7460289\tNutrition Examination Survey\n",
      "7813562\tENA\n",
      "9902798\tMedicare\n",
      "7806138\tENA\n",
      "7451075\tMIMIC\n",
      "8313002\tMedicare\n",
      "7682595\tENA\n",
      "7589079\tGene Expression Omnibus\n",
      "7337779\tData Warehouse\n",
      "8879908\tData Warehouse\n",
      "8138031\tGISAID\n",
      "8079857\tMedicaid\n",
      "8691345\tMedicaid\n",
      "7518800\tGEO\n",
      "7281497\tGISAID\n",
      "8826518\tGISAID\n",
      "8929045\tGISAID\n",
      "8014608\tSRA\n",
      "8176267\tGISAID\n",
      "8906890\tMedicaid\n",
      "8877338\tNextstrain\n",
      "8800303\tENA\n",
      "7718781\tMedicare\n",
      "9088528\tGISAID\n",
      "8977907\tMedicare\n",
      "7927505\tSRA\n",
      "7959679\tGene Expression Omnibus\n",
      "7720151\tGISAID\n",
      "8242375\tGISAID\n",
      "7436652\tData Warehouse\n",
      "7706937\tGISAID\n",
      "7598682\tMedicare\n",
      "9017565\tGISAID\n",
      "9028729\tGEO\n",
      "9015234\tGISAID\n",
      "8685184\tGISAID\n",
      "9671555\tGEO\n",
      "8204200\tMedicare\n",
      "7409725\tGISAID\n",
      "7832442\tGISAID\n",
      "8166511\tENA\n",
      "8267026\tMedicare\n",
      "9045195\tGISAID\n",
      "8271325\tHost Genetics Initiative\n",
      "8233849\tGISAID\n",
      "7665524\tData Warehouse\n",
      "7998535\tENA\n",
      "8143327\tGISAID\n",
      "8800500\tData Warehouse\n",
      "7294288\tOptumLabs\n",
      "7457216\tGISAID\n",
      "8106515\tData Warehouse\n",
      "7722495\tGEO\n",
      "7821175\tData Warehouse\n",
      "8263094\tMedicare\n",
      "9088576\tnextstrain\n",
      "8960748\tMedicaid\n",
      "8733510\tGISAID\n",
      "8788445\tGISAID\n",
      "8242629\tMedicare\n",
      "8242901\tCMS\n",
      "7442128\tGISAID\n",
      "9000049\tMedicaid\n",
      "8865066\tGISAID\n",
      "8213524\tSRA\n",
      "8531388\tReactome\n",
      "8064472\tGISAID\n",
      "8503053\tHost Genetics Initiative\n",
      "7532803\tMIMIC\n",
      "7874846\tGISAID\n",
      "8269753\tMedicare\n",
      "8589563\tMedicare\n",
      "8045659\tSRA\n",
      "8294418\tSRA\n",
      "8242767\tCMS\n",
      "8470771\tGISAID\n",
      "8498775\tData Warehouse\n",
      "8211190\tMedicare\n",
      "8988455\tMedicare\n",
      "8242398\tENA\n",
      "8153470\tGISAID\n",
      "7489775\tMedicare\n",
      "7166400\tGISAID\n",
      "8874890\tData Warehouse\n",
      "8533401\tGISAID\n",
      "8089447\tGISAID\n",
      "8657362\tMedicare\n",
      "8254381\tENA\n",
      "7258756\tGISAID\n",
      "8322982\tHost Genetics Initiative\n",
      "7500898\tGene Expression Omnibus\n",
      "7255323\tENA\n",
      "9653061\tMedicaid\n",
      "8426813\tGISAID\n",
      "7510634\tENA\n",
      "8609622\tGEO\n",
      "8500600\tN3C\n",
      "8522424\tGISAID\n",
      "8568320\tGISAID\n",
      "7174922\tGISAID\n",
      "7402108\tCMS\n",
      "8599730\tNextstrain\n",
      "8003345\tGISAID\n",
      "8205126\tENA\n",
      "8317002\tData Warehouse\n",
      "7877199\tMedicare\n",
      "7473213\tGenbank\n",
      "8307803\tGISAID\n",
      "8799442\tGene Expression Omnibus\n",
      "7314204\tGISAID\n",
      "8242717\tCMS\n",
      "8685582\tMedicare\n",
      "7465887\tNHANES\n",
      "8188262\tGISAID\n",
      "8630285\tMedicaid\n",
      "7905334\tMedicare\n",
      "8829845\tOptum\n",
      "8251413\tData Warehouse\n",
      "7326305\tMedicare\n",
      "8056476\tMedicaid\n",
      "8691200\tGISAID\n",
      "7445393\tGenbank\n",
      "8652737\tMedicare\n",
      "7511831\tENA\n",
      "9196692\tN3C\n",
      "8242621\tMedicare\n",
      "9088347\tGenbank\n",
      "7234208\tNational Health Survey\n",
      "7142689\tReactome\n",
      "8286236\tVHA\n",
      "7428802\tData Warehouse\n",
      "9750189\tData Warehouse\n",
      "8421095\tMedicare\n",
      "7418792\tSRA\n",
      "8498984\tMedicaid\n",
      "8457639\tVeterans Health Administration\n",
      "8099703\tReactome\n",
      "8428849\tMedicare\n",
      "7665476\tGenbank\n",
      "8893930\tMedicare\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8914131\tENA\n",
      "8299215\tMedicare\n",
      "7251277\tNextstrain\n",
      "8171993\tSRA\n",
      "7228393\tGISAID\n",
      "8713434\tGenbank\n",
      "8947833\tGene Expression Omnibus\n",
      "8725927\tGISAID\n",
      "8855654\tGISAID\n",
      "8018669\tSRA\n",
      "7463154\tVHA\n",
      "7640577\tNIS\n",
      "9021583\tVeterans Health Administration\n",
      "8943844\tGISAID\n",
      "8447490\tSequence Read Archive\n",
      "7231477\tGISAID\n",
      "7233242\tENA\n",
      "8294486\tData Warehouse\n",
      "8097105\tVHA\n",
      "7890338\tOptum\n",
      "8138329\tCMS\n",
      "8803346\tNextstrain\n",
      "8076216\tGISAID\n",
      "8035651\tGISAID\n",
      "7480334\tSRA\n",
      "8485368\tENA\n",
      "8184080\tGISAID\n",
      "7684019\tMedicare\n",
      "8739765\tMedicare\n",
      "8280612\tGISAID\n",
      "8901586\tSRA\n",
      "9088325\tData Warehouse\n",
      "8762488\tMedicare\n",
      "8013796\tMedicare\n",
      "7684018\tVeterans Health Administration\n",
      "7485934\tMedicare\n",
      "8441748\tENA\n",
      "8349402\tnextstrain\n",
      "8279096\tCDW\n",
      "8880348\tNIS\n",
      "7949745\tGISAID\n",
      "8479448\tMedicare\n",
      "8102052\tMedicare\n",
      "8809573\tMedicare\n",
      "7832795\tGISAID\n",
      "8224693\tMedicare\n",
      "8104036\tMedicaid\n",
      "7971693\tGISAID\n",
      "8908924\tNutrition Examination Survey\n",
      "7251363\tMedicare\n",
      "8233546\tMedicaid\n",
      "8547852\tGISAID\n",
      "7437610\tCMS\n",
      "7436700\tGISAID\n",
      "8661962\tGISAID\n",
      "7917268\tData Warehouse\n",
      "8895417\tENA\n"
     ]
    }
   ],
   "source": [
    "matched_articles = []\n",
    "for i, article in enumerate(articles):\n",
    "    text = article['main text']\n",
    "    rwd_match = rwd_sec_query.search(text)\n",
    "    omics_match = omics_sec_query.search(text)\n",
    "    if rwd_match:\n",
    "        print(\"%s\\t%s\" % (article['pmcid'], rwd_match.group(0)))\n",
    "        matched_articles.append(article)\n",
    "    elif omics_match:\n",
    "        print(\"%s\\t%s\" % (article['pmcid'], omics_match.group(0)))\n",
    "    else:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7b785871",
   "metadata": {},
   "outputs": [],
   "source": [
    "txt_directory=\"./methods_txt/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ed0352c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: it was not able to read a path, a file-like object, or a string as an XML\n"
     ]
    },
    {
     "ename": "XMLSyntaxError",
     "evalue": "Start tag expected, '<' not found, line 1, column 1 (<string>, line 1)",
     "output_type": "error",
     "traceback": [
      "Traceback \u001b[0;36m(most recent call last)\u001b[0m:\n",
      "  File \u001b[1;32m\"/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages/pubmed_parser/utils.py\"\u001b[0m, line \u001b[1;32m31\u001b[0m, in \u001b[1;35mread_xml\u001b[0m\n    tree = etree.parse(path)\n",
      "  File \u001b[1;32m\"src/lxml/etree.pyx\"\u001b[0m, line \u001b[1;32m3521\u001b[0m, in \u001b[1;35mlxml.etree.parse\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m1859\u001b[0m, in \u001b[1;35mlxml.etree._parseDocument\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m1885\u001b[0m, in \u001b[1;35mlxml.etree._parseDocumentFromURL\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m1789\u001b[0m, in \u001b[1;35mlxml.etree._parseDocFromFile\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m1177\u001b[0m, in \u001b[1;35mlxml.etree._BaseParser._parseDocFromFile\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m615\u001b[0m, in \u001b[1;35mlxml.etree._ParserContext._handleParseResultDoc\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m725\u001b[0m, in \u001b[1;35mlxml.etree._handleParseResult\u001b[0m\n",
      "\u001b[0;36m  File \u001b[0;32m\"src/lxml/parser.pxi\"\u001b[0;36m, line \u001b[0;32m652\u001b[0;36m, in \u001b[0;35mlxml.etree._raiseParseError\u001b[0;36m\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m\u001b[0;31m:\u001b[0m Error reading file './litcovid_pmc_fulltext/xml/PMC8598683.xml': failed to load external entity \"./litcovid_pmc_fulltext/xml/PMC8598683.xml\"\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "Traceback \u001b[0;36m(most recent call last)\u001b[0m:\n",
      "  File \u001b[1;32m\"/usr/local/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py\"\u001b[0m, line \u001b[1;32m3444\u001b[0m, in \u001b[1;35mrun_code\u001b[0m\n    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \u001b[1;32m\"/tmp/ipykernel_56294/3437366670.py\"\u001b[0m, line \u001b[1;32m4\u001b[0m, in \u001b[1;35m<module>\u001b[0m\n    metadata = pp.parse_pubmed_xml(xml_path)\n",
      "  File \u001b[1;32m\"/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages/pubmed_parser/pubmed_oa_parser.py\"\u001b[0m, line \u001b[1;32m128\u001b[0m, in \u001b[1;35mparse_pubmed_xml\u001b[0m\n    tree = read_xml(path, nxml)\n",
      "  File \u001b[1;32m\"/data/xzuo/.conda/envs/dissertation/lib/python3.8/site-packages/pubmed_parser/utils.py\"\u001b[0m, line \u001b[1;32m36\u001b[0m, in \u001b[1;35mread_xml\u001b[0m\n    tree = etree.fromstring(path)\n",
      "  File \u001b[1;32m\"src/lxml/etree.pyx\"\u001b[0m, line \u001b[1;32m3237\u001b[0m, in \u001b[1;35mlxml.etree.fromstring\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m1896\u001b[0m, in \u001b[1;35mlxml.etree._parseMemoryDocument\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m1777\u001b[0m, in \u001b[1;35mlxml.etree._parseDoc\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m1082\u001b[0m, in \u001b[1;35mlxml.etree._BaseParser._parseUnicodeDoc\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m615\u001b[0m, in \u001b[1;35mlxml.etree._ParserContext._handleParseResultDoc\u001b[0m\n",
      "  File \u001b[1;32m\"src/lxml/parser.pxi\"\u001b[0m, line \u001b[1;32m725\u001b[0m, in \u001b[1;35mlxml.etree._handleParseResult\u001b[0m\n",
      "\u001b[0;36m  File \u001b[0;32m\"src/lxml/parser.pxi\"\u001b[0;36m, line \u001b[0;32m654\u001b[0;36m, in \u001b[0;35mlxml.etree._raiseParseError\u001b[0;36m\u001b[0m\n",
      "\u001b[0;36m  File \u001b[0;32m\"<string>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31mXMLSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m Start tag expected, '<' not found, line 1, column 1\n"
     ]
    }
   ],
   "source": [
    "for i, article in enumerate(matched_articles):\n",
    "    xml_name = \"PMC{}.xml\".format(article['pmcid'])\n",
    "    xml_path = os.path.join(download_directory_path, xml_name)\n",
    "    metadata = pp.parse_pubmed_xml(xml_path)\n",
    "#     try:\n",
    "#         full_text = pp.parse_pubmed_paragraph(pmc_xml_path, all_paragraph=True)\n",
    "#     except:\n",
    "#         print(\"Error in %s.\" % xml_name)\n",
    "#         os.remove(xml_path)\n",
    "    full_text = pp.parse_pubmed_paragraph(pmc_xml_path, all_paragraph=True)\n",
    "    paras = []\n",
    "    for j, para in enumerate(full_text):\n",
    "        if section_pattern.search(para['section']):\n",
    "            paras.append(para['text'])\n",
    "        else:\n",
    "            continue\n",
    "    main_text = \"\\n\".join(paras)\n",
    "    \n",
    "    # Format lines.\n",
    "    lines = all_text_per_article.splitlines()\n",
    "    lines = [line for line in lines if len(line) > 2]\n",
    "    new_lines = []\n",
    "    for line in lines:\n",
    "        if not new_lines or not(line.startswith((\")\", \"-\", \",\"))):\n",
    "            new_lines.append(line)\n",
    "        else:\n",
    "            new_lines[-1] += line\n",
    "    new_text_per_article = \"\\n\\n\".join(new_lines)\n",
    "    # print(all_text_per_article)\n",
    "    \n",
    "    txt_name = xml_name.replace(\".xml\", \".txt\")\n",
    "    output_path = os.path.join(txt_directory, txt_name)\n",
    "    with open(output_path, \"w\", encoding='utf-8') as output_file:\n",
    "        output_file.write(new_text_per_article)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3d1278",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
